File: fix-ts-errors.js
#!/usr/bin/env node
/**
 * Script to automatically fix common TypeScript errors in test files
 */

const fs = require('fs');
const path = require('path');

function fixFile(filePath) {
  let content = fs.readFileSync(filePath, 'utf8');
  let modified = false;

  // Fix 1: Wrap arrow functions that return numbers in void context
  // Pattern: (cell) => array.push(cell) -> (cell) => { array.push(cell); }
  const pushPattern = /\(([\w]+)\)\s*=>\s*([\w]+)\.push\(\1\)/g;
  const newContent = content.replace(pushPattern, (match, param, array) => {
    modified = true;
    return `(${param}) => {\n        ${array}.push(${param});\n      }`;
  });

  if (modified) {
    content = newContent;
  }

  // Fix 2: Add optional chaining for potentially undefined array access
  // Pattern: array[0].property -> array[0]?.property
  // This is more complex and needs careful handling

  if (modified) {
    fs.writeFileSync(filePath, content, 'utf8');
    console.log(`Fixed: ${filePath}`);
    return true;
  }

  return false;
}

// Get all test files
const testDirs = [
  'src/circulatory/__tests__',
  'src/immune/__tests__',
  'src/muscular/__tests__',
  'src/respiratory/__tests__',
  'src/skeletal/__tests__',
  'src/tools/__tests__',
  'src/visualization/__tests__',
];

let totalFixed = 0;
testDirs.forEach(dir => {
  const fullPath = path.join(__dirname, dir);
  if (fs.existsSync(fullPath)) {
    const files = fs.readdirSync(fullPath).filter(f => f.endsWith('.test.ts'));
    files.forEach(file => {
      const filePath = path.join(fullPath, file);
      if (fixFile(filePath)) {
        totalFixed++;
      }
    });
  }
});

console.log(`\nTotal files fixed: ${totalFixed}`);

File: fix_ts_errors.py
#!/usr/bin/env python3
"""
Script to fix common TypeScript errors in test files
"""

import os
import re
from pathlib import Path

def fix_push_callbacks(content):
    """Fix (param) => array.push(param) to (param) => { array.push(param); }"""
    # Pattern: (word) => word.push(word)
    pattern = r'\((\w+)\)\s*=>\s*(\w+)\.push\(\1\)'
    replacement = r'(\1) => {\n        \2.push(\1);\n      }'
    return re.sub(pattern, replacement, content)

def fix_increment_callbacks(content):
    """Fix () => count++ to () => { count++; }"""
    pattern = r'\(\)\s*=>\s*(\w+)\+\+'
    replacement = r'() => {\n        \1++;\n      }'
    return re.sub(pattern, replacement, content)

def fix_optional_access(content):
    """Fix array[0].property to array[0]?.property"""
    # This is more complex - look for specific patterns
    replacements = [
        (r'(\w+)\[0\]\.payload', r'\1[0]?.payload'),
        (r'(\w+)\[0\]\.message', r'\1[0]?.message'),
        (r'(\w+)\[0\]\.id', r'\1[0]?.id'),
        (r'(\w+)\[1\]\.payload', r'\1[1]?.payload'),
        (r'(\w+)\[1\]\.message', r'\1[1]?.message'),
    ]

    for pattern, replacement in replacements:
        content = re.sub(pattern, replacement, content)

    return content

def fix_file(filepath):
    """Fix a single file"""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    original = content

    # Apply fixes
    content = fix_push_callbacks(content)
    content = fix_increment_callbacks(content)

    # Only write if changed
    if content != original:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        return True

    return False

def main():
    """Main function"""
    src_dir = Path('src')
    files_fixed = 0

    # Find all test and story files
    for filepath in src_dir.rglob('*.test.ts'):
        if fix_file(filepath):
            print(f'Fixed: {filepath}')
            files_fixed += 1

    for filepath in src_dir.rglob('*.stories.ts'):
        if fix_file(filepath):
            print(f'Fixed: {filepath}')
            files_fixed += 1

    print(f'\nTotal files fixed: {files_fixed}')

if __name__ == '__main__':
    main()
